File: Review-06-CPP-Shapes/Example-8/source/runShapes.cpp
// Thomas Kennedy
// Review Example: Inheritance and the Factory Model

#include <iostream>
#include <iomanip>
#include <fstream>
#include <vector>

#include <memory>
#include <iterator>
#include <algorithm>

#include "utilities.h"
#include "shapeFactory.h"
// #include "ShapeCollection.h" // No longer needed

/** @file */

using namespace std;
using namespace utilities;

const string PROGRAM_HEADING[] = {
    "Objects & Inheritance: 2-D Shapes",
    "Thomas J. Kennedy"
};  ///< Program Title

const int HEADING_LINES = 2;  ///< Number of lines in Program Heading

using ShapeCollection = std::vector<std::unique_ptr<Shape>>;

/**
 * Read shapes from an input stream
 * and construct a `ShapeCollection` object
 */
ShapeCollection readShapes(std::istream& ins);

/**
 * Print shapes from a `ShapeCollection` to a
 * specified output stream
 */
void printShapes(std::ostream& outs, const ShapeCollection& toPrint);

/**
 * Print shape names for all `Shape`s in a `ShapeCollection` to a
 * specified output stream
 */
void printShapeNames(std::ostream& outs, const ShapeCollection& toPrint);

/**
 * This program accepts command line
 * arguments
 */
int main(int argc, char** argv)
{
    if (argc < 2) {
        cout << "No input file provided." << "\n";
        cout << "Usage: " << argv[0] << " input_file" << "\n";
        return 1;
    }

    ifstream shapesFile(argv[1]);

    if (!shapesFile) {
        cout << "Error: " << argv[1] << "could not be opened" << "\n";
        return 2;
    }

    // Set formatting
    cout.precision(4);
    cout.setf(ios::fixed);

    printProjectHeading(cout, PROGRAM_HEADING, HEADING_LINES);

    // Examine the ShapeFactory
    printHeading(cout, "Available Shapes", 38, '~');

    // List the available shapes
    ShapeFactory::listKnown(cout);
    printHorizontalLine(cout, '-', 38);
    cout << right << setw(2)
         << ShapeFactory::numberKnown()
         << " shapes available."
         << "\n";

    cout << "\n";

    // Create 5 "Random" Shapes
    ShapeCollection shapes = readShapes(shapesFile);

    // Print all the shapes
    printHeading(cout, "Display All Shapes", 38, '~');
    // cout << shapes;
    printShapes(cout, shapes);

    printHeading(cout, "Display Shape Names", 38, '~');
    // cout << shapes;
    printShapeNames(cout, shapes);

    printHeading(cout, "Display Largest Shape (Area)", 38, '~');
    ShapeCollection::const_iterator it;

    it = std::max_element(shapes.begin(), shapes.end(),
                          // [](const unique_ptr<Shape>& lhs, const unique_ptr<Shape>& rhs) {
                          [](const auto& lhs, const auto& rhs) {
                              return (lhs)->area() < (rhs)->area();
                          });
    cout << *(*it) << "\n";

    printHeading(cout, "Display Smallest Shape (Perimeter)", 38, '~');
    it = std::min_element(shapes.begin(), shapes.end(),
                          [](const auto& lhs, const auto& rhs) {
                              return (lhs)->perimeter() < (rhs)->perimeter();
                          });
    cout << *(*it) << "\n";

    printHeading(cout, "Display Sorted by Name", 38, '~');
    std::sort(shapes.begin(), shapes.end(),
              [](const auto& lhs, const auto& rhs) {
                  return (lhs)->name() < (rhs)->name();
              });
    printShapes(cout, shapes);

    return 0;
}

//------------------------------------------------------------------------------
ShapeCollection readShapes(std::istream& ins)
{
    ShapeCollection collection;
    /*
    Shape*          s;

    ins >> ws;

    while (ins >> s) {
        if (s != nullptr) {
            collection.emplace_back(s);
        }
        ins >> ws;
    }
    */
    // cout << collection << "\n";  // Not legal anymore

    std::istream_iterator<Shape*> it(ins);
    std::istream_iterator<Shape*> end;
    // Let us re-enact the Back-to-the-Future I Guitar Scene!
    for_each(it, end,
             [&collection](Shape* s) {
                  if (s != nullptr) {
                      collection.emplace_back(s);
                  }
             });

    return collection;
}

//------------------------------------------------------------------------------
void printShapes(std::ostream& outs, const ShapeCollection& toPrint)
{
    // for (const Shape* s : toPrint) {
    for (const std::unique_ptr<Shape>& s : toPrint) {
        // outs << s << "\n";  // oops
        outs << *s << "\n";
    }
}

//------------------------------------------------------------------------------
void printShapeNames(std::ostream& outs, const ShapeCollection& toPrint)
{
    for (const std::unique_ptr<Shape>& s : toPrint) {
        outs << s->name() << "\n";
    }
}

File: 06-Java-Shapes/Example-4/src/main/java/RunShapes.java
//Programmer : Thomas Kennedy

import java.util.List;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.Scanner;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileNotFoundException;

import utilities.Utilities;
import static utilities.Utilities.projectHeading;
import static utilities.Utilities.heading;
import static utilities.Utilities.projectHeading;
import static utilities.Utilities.horizontalLine;

import shapes.*;


/**
 * This is the Java version of the previous C++ Shapes Inheritance Example.
 *
 * @author Thomas J Kennedy
 */
public class RunShapes {
    /**
     * This is the Program Title.
     */
    private static final String[] PROGRAM_HEADING = {
        "Objects & Inheritance: 2-D Shapes",
        "Thomas J. Kennedy"
    };

    /**
     * The default heading width.
     */
    private static final int H_WIDTH = 38;

    /**
     * This is the main function.
     *
     * @param args args[0] must specify the filename of a valid shape text file
     *
     * @throws CloneNotSupportedException if a Shape subclass can
     *     not be copied.
     */
    public static void main(String[] args)
        throws CloneNotSupportedException
    {
        // Command line argument and File validation
        BufferedReader shapesFile = null;
        try {
            shapesFile = new BufferedReader(new FileReader(args[0]));
        }
        catch (ArrayIndexOutOfBoundsException e) {
            System.out.println("Usage: java -jar {jarfile} {inputTextFile}");
            System.exit(1);
        }
        catch (FileNotFoundException e) {
            System.out.println("File (" + args[0] + ") could not be opened.");
            System.exit(2);
        }

        // Print main program heading
        System.out.println(projectHeading(PROGRAM_HEADING, Utilities.W_WIDTH));

        //----------------------------------------------------------------------
        // Examine the ShapeFactory
        //----------------------------------------------------------------------
        System.out.println(heading("Available Shapes", H_WIDTH, '*'));
        System.out.print(ShapeFactory.listKnown());
        System.out.println(horizontalLine('-', H_WIDTH));
        System.out.printf("%2d shapes available.%n", ShapeFactory.numberKnown());
        System.out.println();

        List<Shape> shapes = readShapes(shapesFile);

        // Print all the shapes
        System.out.println(heading("Display All Shapes", H_WIDTH, '*'));
        printShapes(shapes);

        // Using an iterator
        System.out.println(heading("Display Shape Names", 38, '~'));
        printShapeNames(shapes);
        System.out.println();

        System.out.println(heading("Display Largest Shape (Area)", 38, '~'));
        Shape largestShape = findLargestShapeByArea(shapes);
        System.out.println(largestShape);
    }

    /**
     * Read shapes from an input stream
     * and construct an `ArrayList<Shape>` object.
     *
     * @param shapesFile input source
     *
     * @return collection of read-in shapes
     *
     * @throws CloneNotSupportedException if the `ShapeFactory` fails to clone a
     *     model shape
     */
    private static List<Shape> readShapes(BufferedReader shapesFile)
        throws CloneNotSupportedException
    {
        Scanner scanner = new Scanner(shapesFile);
        ArrayList<Shape> collection = new ArrayList<Shape>();

        while (scanner.hasNextLine()) {
            String line   = scanner.nextLine();

            int    sIndex = line.indexOf(';', 0);
            String name   = line.substring(0, sIndex); // [0, sIndex)

            Scanner lineScanner = new Scanner(line.substring(sIndex + 1,
                                                             line.length()));
            // System.out.println(line);
            // System.out.println(name);

            Shape s = ShapeFactory.createShape(name);

            if (s != null) {
                s.read(lineScanner);
                collection.add(s);
            }
            //else {
                // Ignore inLineScanner and line
                // they will cease to exist
                // after this loop iteration
            //}
        }

        return collection;
    }

    /**
     * Print shapes from a `ArrayList<Shape>`
     */
    private static void printShapes(List<Shape> toPrint)
    {
        for (Shape s : toPrint) {
            System.out.println(s);
        }
    }

    /**
     * Print shape names for all `Shape`s in a `ArrayList<Shape>` to a
     * specified output stream
     */
    private static void printShapeNames(List<Shape> toPrint)
    {
        // C++ Container<Shape*>::iterator it = shapes.begin()
        Iterator<Shape> it = toPrint.iterator();

        // C++ while (it != toPrint.end())
        while (it.hasNext()) {
            Shape s = it.next();
            System.out.println(s.name());
        }
    }

    /**
     * Find the largest `Shape` (by area) in a `ArrayList<Shape>`
     *
     * @return an iterator at the position of the largest `Shape`
     */
    private static Shape findLargestShapeByArea(List<Shape> collection)
    {
        Iterator<Shape> it = collection.iterator();

        // The first shape is the largest
        // until I look at more
        Shape largestShape = it.next();

        while (it.hasNext()) {
            Shape s = it.next();

            if (s.area() > largestShape.area()) {
                largestShape = s;
            }
        }

        return largestShape;
    }
}
